### PR TITLE
Add cherry as alternative CLJS evaluator

### PR DESCRIPTION
**Pull request by borkdude**

### CODE REVIEW

**Comment by borkdude** [comment]
@ikitommi I'm sure you have some feedback, fire away. I think I should add some docs on how to enable cherry with malli. In short: just add the cherry dependency and require `malli.cherry`.

File: src/malli/cherry.cljs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  ([code] (eval code nil))
+  ([code _opts]
+   (cond (or (symbol? code)
+             (seq? code))
+         (cherry/eval-form code cherry-opts)
+         (string? code)
+         (cherry/eval-string code cherry-opts)
+         :else code)))
+
+(set! malli.core/eval eval)
```
Comment (ikitommi) [inline src/malli/cherry.cljs:23]: You can swap the var like this in cljs, I did not know, never done it. Could use this other things too and maybe swap the sci to work like this. Thanks for showing.

**Review by ikitommi — APPROVED** [review]
Looks good 👍

**Comment by borkdude** [comment]
A small additional note:

To use SCI + cherry in one advanced build you need to use the newest commit from CLJS:

https://github.com/clojure/clojurescript/commit/0ccd19fc79f04a5082764901bd6413a5841ca428

**Comment by ikitommi** [comment]
Needed 721eb361b5064ffa6b7fd4f072a27baed8096e3e so that the bundle size test app still work, e.g.

```bash
npx shadow-cljs run shadow.cljs.build-report app /tmp/report.html
npx shadow-cljs run shadow.cljs.build-report app-sci /tmp/report.html #failed

npx shadow-cljs run shadow.cljs.build-report app2 /tmp/report.html
npx shadow-cljs run shadow.cljs.build-report app2-sci /tmp/report.html #failed
```

**Comment by ikitommi** [comment]
Happy to merge this already, but if you could write a quick how-to into README, that would be awesome.

Maybe add shadow `:build` for cherry too to see the effects on bundle size?

**Comment by borkdude** [comment]
@ikitommi Done

```
| Resource                                    |   Optimized |  Total % |          JS |      Source |
|---------------------------------------------+-------------+----------+-------------+-------------|
| jar | cljs/core.cljs                        |    367,4 KB |  43,18 % |      1,3 MB |    348,0 KB |
|     | malli/core.cljc                       |    167,6 KB |  19,69 % |    631,7 KB |    119,2 KB |
|     | cherry/internal/macros.cljc           |     45,0 KB |   5,28 % |    134,0 KB |     21,0 KB |
|     | cherry/compiler.cljc                  |     41,9 KB |   4,92 % |     83,8 KB |     18,9 KB |
|     | malli/cherry.cljs                     |     37,3 KB |   4,39 % |    129,9 KB |    694,0 B  |
| jar | edamame/impl/parser.cljc              |     36,1 KB |   4,24 % |    149,1 KB |     35,2 KB |
|     | squint/compiler_common.cljc           |     23,7 KB |   2,79 % |     92,6 KB |     28,1 KB |
|     | cherry/internal/fn.cljc               |     21,6 KB |   2,54 % |     61,6 KB |     15,8 KB |
|     | cherry/internal/protocols.cljc        |     21,1 KB |   2,48 % |     64,0 KB |     18,0 KB |
|     | malli/impl/regex.cljc                 |     20,1 KB |   2,36 % |     97,6 KB |     26,4 KB |
|     | cherry/internal/deftype.cljc          |      9,0 KB |   1,06 % |     22,8 KB |      7,1 KB |
| jar | cljs/tools/reader/edn.cljs            |      6,8 KB |   0,80 % |     34,9 KB |     15,4 KB |
| jar | borkdude/dynaload.cljc                |      6,8 KB |   0,79 % |     15,6 KB |      8,8 KB |
| jar | cljs/tools/reader/impl/errors.cljs    |      4,2 KB |   0,50 % |     25,1 KB |      6,9 KB |
|     | cherry/internal/destructure.cljc      |      4,2 KB |   0,50 % |     12,1 KB |      7,5 KB |
| jar | cljs/tools/reader/impl/utils.cljs     |      4,0 KB |   0,46 % |     19,2 KB |      2,7 KB |
| jar | clojure/string.cljs                   |      3,9 KB |   0,46 % |     14,7 KB |      8,4 KB |
| jar | cljs/tools/reader/reader_types.cljs   |      3,6 KB |   0,43 % |     38,8 KB |      8,9 KB |
| jar | cljs/reader.cljs                      |      3,5 KB |   0,42 % |     19,5 KB |      7,2 KB |
| jar | edamame/impl/syntax_quote.cljc        |      3,2 KB |   0,37 % |      9,2 KB |      4,6 KB |
| jar | cljs/tools/reader/impl/commons.cljs   |      2,5 KB |   0,29 % |      9,4 KB |      4,4 KB |
| jar | edamame/impl/ns_parser.cljc           |      1,8 KB |   0,21 % |      6,9 KB |      2,4 KB |
| jar | cljs/tools/reader/impl/inspect.cljs   |      1,7 KB |   0,20 % |      8,2 KB |      2,7 KB |
|     | malli/registry.cljc                   |      1,6 KB |   0,19 % |     24,7 KB |      2,8 KB |
| jar | edamame/impl/read_fn.cljc             |      1,6 KB |   0,19 % |      5,5 KB |      2,2 KB |
| jar | cljs/tagged_literals.cljc             |      1,6 KB |   0,19 % |      4,7 KB |      3,0 KB |
|     | malli/sci.cljc                        |      1,5 KB |   0,17 % |      3,9 KB |    795,0 B  |
|     | cherry/internal/loop.cljc             |      1,2 KB |   0,15 % |      4,5 KB |      1,7 KB |
|     | malli/impl/util.cljc                  |      1,2 KB |   0,14 % |      4,5 KB |      2,9 KB |
| jar | goog/string/stringformat.js           |      1,1 KB |   0,13 % |      7,8 KB |      7,8 KB |
| jar | cljs/tools/reader.cljs                |    932,0 B  |   0,11 % |     83,2 KB |     34,4 KB |
| jar | goog/object/object.js                 |    522,0 B  |   0,06 % |     21,5 KB |     21,5 KB |
| jar | goog/array/array.js                   |    489,0 B  |   0,06 % |     61,5 KB |     61,5 KB |
| jar | goog/string/stringbuffer.js           |    400,0 B  |   0,05 % |      2,3 KB |      2,3 KB |
|     | cherry/embed.cljs                     |    378,0 B  |   0,04 % |      5,3 KB |      1,4 KB |
| jar | goog/string/string.js                 |    360,0 B  |   0,04 % |     47,1 KB |     47,1 KB |
| jar | goog/string/internal.js               |    348,0 B  |   0,04 % |     12,2 KB |     12,2 KB |
| jar | goog/base.js                          |    313,0 B  |   0,04 % |    115,2 KB |    115,2 KB |
|     | malli/app.cljc                        |    189,0 B  |   0,02 % |    685,0 B  |    300,0 B  |
| jar | edamame/core.cljc                     |    124,0 B  |   0,01 % |      9,1 KB |      4,7 KB |
```
[report.html.txt](https://github.com/metosin/malli/files/11201882/report.html.txt)

**Pull request merged**